go/proxyd/daisychain.go

```diff
@@ -0,0 +1,528 @@
+	epoch5RPCURL := config.Backends.Epoch5RPCURL
+	epoch6RPCURL := config.Backends.Epoch6RPCURL
+
+	urls := []string{
+		epoch1RPCURL,
+		epoch2RPCURL,
+		epoch3RPCURL,
+		epoch4RPCURL,
+		epoch5RPCURL,
+		epoch6RPCURL,
+	}
+
+	defined := false
+	for i, url := range urls {
+		if url != "" {
+			log.Info("epoch rpc url defined", "epoch", i+1)
+			defined = true
+		}
+	}
+	if !defined {
+		panic("must define one epoch url")
+	}
+
+	// parse the config
+	srv := NewDaisyChainServer(urls)
+
+	if config.Metrics.Enabled {
+		addr := fmt.Sprintf("%s:%d", config.Metrics.Host, config.Metrics.Port)
+		log.Info("starting metrics server", "addr", addr)
+		go http.ListenAndServe(addr, promhttp.Handler())
+	}
+
+	// To allow integration tests to cleanly come up, wait
+	// 10ms to give the below goroutines enough time to
+	// encounter an error creating their servers
+	errTimer := time.NewTimer(10 * time.Millisecond)
+
+	if config.Server.RPCPort != 0 {
+		go func() {
+			if err := srv.RPCListenAndServe(config.Server.RPCHost, config.Server.RPCPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("RPC server shut down")
+					return
+				}
+				log.Crit("error starting RPC server", "err", err)
+			}
+		}()
+	}
+
+	if config.Server.WSPort != 0 {
+		go func() {
+			if err := srv.WSListenAndServe(config.Server.WSHost, config.Server.WSPort); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					log.Info("WS server shut down")
+					return
+				}
+				log.Crit("error starting WS server", "err", err)
+			}
+		}()
+	}
+
+	<-errTimer.C
+	log.Info("started daisychain")
+
+	return func() {
+		log.Info("shutting down daisychain")
+		srv.Shutdown()
+		log.Info("goodbye")
+	}, nil
+}
+
+// TODO: batch support
+func (s *DaisyChainServer) HandleRPC(w http.ResponseWriter, r *http.Request) {
+	ctx := s.populateContext(w, r)
+	if ctx == nil {
+		return
+	}
+
+	log.Info(
+		"received RPC request",
+		"req_id", GetReqID(ctx),
+		"auth", GetAuthCtx(ctx),
+		"user_agent", r.Header.Get("user-agent"),
+	)
+
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, s.maxBodySize))
+	if err != nil {
+		log.Error("error reading request body", "err", err)
+		writeRPCError(ctx, w, nil, ErrInternal)
+		return
+	}
+	RecordRequestPayloadSize(ctx, len(body))
+
+	req, err := ParseRPCReq(body)
+	if err != nil {
+		log.Info("error parsing RPC call", "source", "rpc", "err", err)
+		writeRPCError(ctx, w, nil, err)
+		return
+	}
+
+	// See if the incoming method is in the list of methods
+	// that need to be proxied by observing the request options
+	argType, ok := argTypes[req.Method]
+	if !ok {
+		// The request can be forwarded to the most recent node.
+		// TODO: finalize this functionality, because it could
+		// be forward to one of the two most recent nodes, if the
+		// most recent node doesn't start with a 0 blocknumber
+		// We haven't completely derisked the diff to geth it
+		// is to have non contiguous block data
+		if s.epoch6RPCURL == "" {
```

Comment:
Would it make sense to have a configurable fallback backend rather than always use the last?

Reply:
Something that is important as a design goal here is that it is very explicit  with the config - this is because we don't want users to be thinking they are sending requests to a node when they are actually sending to a different node